### PR TITLE
create all tmpdirs inside build_dir

### PR DESCRIPTION
**Pull request by hjmodi**
* CLOUDBLD-7743

Signed-off-by: Harsh Modi <hmodi@redhat.com>

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] JSON/YAML configuration changes are updated in the relevant schema
- [ ] Changes to metadata also update the documentation for the metadata
- [ ] Pull request has a link to an osbs-docs PR for user documentation updates
- [ ] New feature can be disabled from a configuration file

### CODE REVIEW

**Review by rcerven — DISMISSED** [review]

**Comment by chmeliik** [comment]
Noticed one more: https://github.com/containerbuildsystem/atomic-reactor/blob/eb2f4efe480d458ccd663282d8f369636de871c6/atomic_reactor/plugins/post_rpmqa.py#L93

What about `NamedTemporaryFile`s? I guess those can just stay in /tmp?

**Comment by rcerven** [comment]
> Noticed one more:
> 
> https://github.com/containerbuildsystem/atomic-reactor/blob/eb2f4efe480d458ccd663282d8f369636de871c6/atomic_reactor/plugins/post_rpmqa.py#L93
> 
> What about `NamedTemporaryFile`s? I guess those can just stay in /tmp?

mmm actually there is also plugins/post_rpmqa.py:        with tempfile.TemporaryDirectory() as rpmdb_dir:

I'd say all should be in that workdir

**Comment by chmeliik** [comment]
Just throwing this idea out there - would it be better to put all the temporary directories and files in a "temp" subdirectory of the build_dir rather than directly in the build_dir root?

The build_dir already contains the cloned repo and I'm also planning to keep all the per-architecture build directories in it. While naming conflicts between tempfiles and per-arch build dirs are very unlikely, I would still feel calmer if I knew everything temporary is hidden away in a subdir

**Comment by rcerven** [comment]
> Just throwing this idea out there - would it be better to put all the temporary directories and files in a "temp" subdirectory of the build_dir rather than directly in the build_dir root?
> 
> The build_dir already contains the cloned repo and I'm also planning to keep all the per-architecture build directories in it. While naming conflicts between tempfiles and per-arch build dirs are very unlikely, I would still feel calmer if I knew everything temporary is hidden away in a subdir

sounds good to me

**Comment by ben-alkov** [comment]
@hjmodi; I see `os.path.join(workflow.source.workdir, 'temp')` a **lot** - couldn't it be stashed in the workflow somehow (e.g. `workflow.tmp_dir` or something)?

File: atomic_reactor/plugins/build_source_container.py [hunk 1]
```diff
@@ -27,7 +27,10 @@ class SourceContainerPlugin(BuildStepPlugin):
     key = PLUGIN_SOURCE_CONTAINER_KEY
 
     def export_image(self, image_output_dir):
-        output_path = os.path.join(tempfile.mkdtemp(), EXPORTED_SQUASHED_IMAGE_NAME)
+        os.makedirs(os.path.join(self.workflow.source.workdir, 'temp'), exist_ok=True)
+        output_path = os.path.join(tempfile.mkdtemp(dir=os.path.join(self.workflow.source.workdir,
+                                                                     'temp')),
+                                   EXPORTED_SQUASHED_IMAGE_NAME)
```
Comment (chmeliik) [inline atomic_reactor/plugins/build_source_container.py:33]: I wouldn't make this a tmpdir, let's just use something like `{workdir}/{EXPORTED_SQUASHED_IMAGE_NAME}`. The output is not temporary, it is the most important output in the source container pipeline (the built image)

File: atomic_reactor/plugins/build_source_container.py [hunk 2]
```diff
@@ -88,7 +91,8 @@ def run(self):
         if maven_source_exists and not os.listdir(maven_source_data_dir):
             self.log.warning("Maven source directory '%s' is empty", maven_source_data_dir)
 
-        image_output_dir = tempfile.mkdtemp()
+        os.makedirs(os.path.join(self.workflow.source.workdir, 'temp'), exist_ok=True)
+        image_output_dir = tempfile.mkdtemp(dir=os.path.join(self.workflow.source.workdir, 'temp'))
```
Comment (chmeliik) [inline atomic_reactor/plugins/build_source_container.py:95]: Same here, I would make this `{workdir}/output/`

File: atomic_reactor/plugins/post_export_operator_manifests.py [hunk 3]
```diff
@@ -81,7 +81,9 @@ def run(self):
         if not self.should_run():
             return
 
-        manifests_archive_dir = tempfile.mkdtemp()
+        os.makedirs(os.path.join(self.workflow.source.workdir, 'temp'), exist_ok=True)
+        manifests_archive_dir = tempfile.mkdtemp(dir=os.path.join(self.workflow.source.workdir,
+                                                                  'temp'))
```
Comment (chmeliik) [inline atomic_reactor/plugins/post_export_operator_manifests.py:86]: Not needed for now, you could leave an OSBS2 TBD comment that this should go in the context dir

File: atomic_reactor/plugins/post_rpmqa.py [hunk 4]
```diff
@@ -90,7 +90,9 @@ def gather_output(self):
             rpmdb_archive.flush()
             tar_archive = tarfile.TarFile(rpmdb_archive.name)
 
-        with tempfile.TemporaryDirectory() as rpmdb_dir:
+        os.makedirs(os.path.join(self.workflow.source.workdir, 'temp'), exist_ok=True)
+        with tempfile.TemporaryDirectory(dir=os.path.join(self.workflow.source.workdir, 'temp')) \
+                as rpmdb_dir:
```
Comment (chmeliik) [inline atomic_reactor/plugins/post_rpmqa.py:95]: Not needed, should stay in /tmp

File: atomic_reactor/plugins/pre_fetch_sources.py [hunk 5]
```diff
@@ -135,7 +135,8 @@ def download_sources(self, sources, insecure=False, download_dir=SRPMS_DOWNLOAD_
         :param download_dir: str, directory where to download content
         :return: str, paths to directory with downloaded sources
         """
-        workdir = tempfile.mkdtemp()
+        os.makedirs(os.path.join(self.workflow.source.workdir, 'temp'), exist_ok=True)
+        workdir = tempfile.mkdtemp(dir=os.path.join(self.workflow.source.workdir, 'temp'))
```
Comment (chmeliik) [inline atomic_reactor/plugins/pre_fetch_sources.py:139]: how about `{workdir}/inputs/` or `{workdir}/sources/` instead?

File: atomic_reactor/plugins/pre_resolve_remote_source.py [hunk 6]
```diff
@@ -222,9 +222,10 @@ def process_request(self, source_request, name, remote_sources_workers_params,
             tarball_filename = REMOTE_SOURCE_TARBALL_FILENAME
             json_filename = REMOTE_SOURCE_JSON_FILENAME
 
+        os.makedirs(os.path.join(self.workflow.source.workdir, 'temp'), exist_ok=True)
         tarball_dest_path = self.cachito_session.download_sources(
             source_request,
-            dest_dir=self.workflow.source.workdir,
+            dest_dir=os.path.join(self.workflow.source.workdir, 'temp'),
```
Comment (chmeliik) [inline atomic_reactor/plugins/pre_resolve_remote_source.py:228]: no change needed

File: atomic_reactor/utils/koji.py [hunk 7]
```diff
@@ -363,7 +363,8 @@ def get_source_tarballs_output(workflow):
 def get_remote_sources_json_output(workflow):
     plugin_results = workflow.prebuild_results.get(PLUGIN_RESOLVE_REMOTE_SOURCE) or {}
     output_json_files = []
-    tmpdir = tempfile.mkdtemp()
+    os.makedirs(os.path.join(workflow.source.workdir, 'temp'), exist_ok=True)
+    tmpdir = tempfile.mkdtemp(dir=os.path.join(workflow.source.workdir, 'temp'))
```
Comment (chmeliik) [inline atomic_reactor/utils/koji.py:367]: Not needed, should stay in /tmp

**Review by chmeliik — COMMENTED** [review]
After understanding the reasoning behind creating tmpdirs inside build_dir a bit better, we have come to the following conclusion
* inputs for the build should just go directly into the build_dir
* temp files that are only used locally by a single plugin can stay in /tmp
* files that are not build inputs and need to be shared between tasks should go somewhere in the context dir
  * not yet, the context sharing needs more design

Left comments accordingly

File: atomic_reactor/plugins/build_source_container.py (hunk 1 above)
Comment (hjmodi) [reply atomic_reactor/plugins/build_source_container.py]: updated.

File: atomic_reactor/plugins/post_export_operator_manifests.py (hunk 3 above)
Comment (hjmodi) [reply atomic_reactor/plugins/post_export_operator_manifests.py]: reverted and added comment

File: atomic_reactor/plugins/build_source_container.py (hunk 2 above)
Comment (hjmodi) [reply atomic_reactor/plugins/build_source_container.py]: updated.

File: atomic_reactor/plugins/post_rpmqa.py (hunk 4 above)
Comment (hjmodi) [reply atomic_reactor/plugins/post_rpmqa.py]: reverted.

File: atomic_reactor/plugins/pre_fetch_sources.py (hunk 5 above)
Comment (hjmodi) [reply atomic_reactor/plugins/pre_fetch_sources.py]: updated to `{workdir}/sources/`

File: atomic_reactor/plugins/pre_resolve_remote_source.py (hunk 6 above)
Comment (hjmodi) [reply atomic_reactor/plugins/pre_resolve_remote_source.py]: reverted.

File: atomic_reactor/utils/koji.py (hunk 7 above)
Comment (hjmodi) [reply atomic_reactor/utils/koji.py]: reverted.

**Comment by hjmodi** [comment]
> @hjmodi; I see `os.path.join(workflow.source.workdir, 'temp')` a **lot** - couldn't it be stashed in the workflow somehow (e.g. `workflow.tmp_dir` or something)?

we're using different names now :D

**Review by rcerven — APPROVED** [review]

File: atomic_reactor/plugins/build_source_container.py [hunk 8]
```diff
@@ -88,7 +87,8 @@ def run(self):
         if maven_source_exists and not os.listdir(maven_source_data_dir):
             self.log.warning("Maven source directory '%s' is empty", maven_source_data_dir)
 
-        image_output_dir = tempfile.mkdtemp()
+        os.makedirs(os.path.join(self.workflow.source.workdir, 'output'), exist_ok=True)
+        image_output_dir = os.path.join(self.workflow.source.workdir, 'output')
```
Comment (mkosiarc) [inline atomic_reactor/plugins/build_source_container.py:91]: Could the order of these 2 be replaced? so we would have:
```
image_output_dir = os.path.join(self.workflow.source.workdir, 'output')
os.makedirs(image_output_dir, exist_ok=True)
```

Comment (hjmodi) [reply atomic_reactor/plugins/build_source_container.py:90]: updated.

**Review by mkosiarc — APPROVED** [review]

**Review by chmeliik — APPROVED** [review]

**Pull request merged**